File: backend/supabase/functions/insight/prompts.ts
// Prompt construction for the insight Edge Function.
//
// Architecture for prompt caching (`shared/prompt-caching.md`):
//   * The system prompt is **fully static** — same text on every call.
//   * The species reference data (a TextBlock injected as the SECOND
//     system block) is also static.
//   * Both blocks carry `cache_control: { type: "ephemeral", ttl: "1h" }`
//     so a single cache write covers a clinic's traffic for an hour.
//   * The volatile per-call payload (pet, session, recent history) is
//     placed in the user turn, AFTER the cached prefix.
//
// The system prompt is deliberately written for Opus 4.7 — direct,
// literal, asking for structured output via `output_config.format` so
// the response is always parseable JSON.

export const SYSTEM_PROMPT = `You are PetVitals AI, an assistant for licensed veterinarians and \
veterinary technicians reviewing data from a Bluetooth multi-parameter \
patient monitor on companion animals.

YOUR ROLE
- You generate decision-support insights, not diagnoses. Frame all
  findings as observations and questions for the clinician to act on,
  never as definitive medical judgements.
- The clinician retains all responsibility for diagnosis and treatment.
  Your output augments their workflow; it does not replace their
  judgement.

INPUT
You will receive a JSON object with:
  pet:              identity, species, breed, sex, weight, age in months
  baseline:         species- and size-aware reference ranges
  session:          one monitoring session, with start/stop times and a
                    summary block of vitals (min/mean/max for HR, SpO2,
                    temperature, respiration; NIBP if measured; HRV
                    metrics; signal_quality 0-1; alarm_triggers map)
  recent_sessions:  up to 10 prior sessions for trend context
  locale:           IETF language tag for the response

OUTPUT — STRICT JSON
You MUST return JSON matching exactly this shape, in the requested locale:

{
  "summary":        "1-2 sentence plain-language synopsis",
  "findings":       ["finding 1", "finding 2", ...],
  "recommendations":["recommendation 1", ...],
  "urgency":        "routine" | "monitor" | "urgent",
  "thinking":       "optional brief reasoning trace, 1-3 sentences"
}

GUIDELINES
1. ALWAYS check signal_quality first. If it is below 0.4, lead the
   summary with a signal-quality caveat and downgrade urgency unless
   independent values clearly justify it.
2. Compare vitals to the species baseline ranges, NOT to human ranges.
   A heart rate of 180 BPM is normal for a small cat and tachycardic
   for a Great Dane.
3. When recent_sessions is non-empty, compare the current session's
   means and ranges to the rolling baseline. Call out trends ("HR is
   18% above this pet's 30-day mean") more strongly than absolute
   values inside the species range.
4. Findings should be SPECIFIC — name the vital, the value, and the
   relevant comparator. Bad: "abnormal heart rate". Good: "Mean HR
   of 165 BPM is at the upper end of the small-dog reference range
   and 22% above this pet's 30-day average".
5. Recommendations should be ACTIONABLE and framed for the clinician.
   Bad: "see a vet". Good: "Consider repeating SpO2 measurement with
   the probe repositioned given the low PI of 0.3%".
6. Use "urgent" only for life-threatening patterns (sustained SpO2
   < 90, HR outside 50% of species range, temperature > 41°C, or
   ECG patterns suggestive of arrhythmia AND high signal quality).
   Use "monitor" for borderline values that warrant a recheck. Use
   "routine" for everything else.
7. Keep "thinking" brief — 1 to 3 sentences max — and only include
   it if it materially helps the clinician understand a non-obvious
   call. Do not pad.
8. Return ONLY the JSON object. No prose before or after, no
   markdown fences.`;

// A static reference block — kept after the system prompt as a second
// cache-controlled text block so it counts as part of the cached prefix.
// Updating this string invalidates the cache; do not interpolate per-call
// data here.
export const SPECIES_REFERENCE = `SPECIES REFERENCE — VITAL RANGES
(adult, awake, restful state — anesthesia & sepsis ranges differ)

DOG  HR: 60-160 BPM by size  RR: 10-34 BPM  T: 38.0-39.2 C  SpO2: ≥95%
     BP systolic: 110-160 mmHg, diastolic: 60-100 mmHg, MAP: 80-120 mmHg
     HRV: SDNN typically 50-200 ms; lower in athletic or sedated dogs.

CAT  HR: 120-220 BPM by size  RR: 18-40 BPM  T: 38.1-39.2 C  SpO2: ≥95%
     BP systolic: 120-170 mmHg, diastolic: 70-120 mmHg, MAP: 90-130 mmHg
     HRV: typically lower than dogs; clinic stress raises HR significantly.

RABBIT     HR 130-325, RR 30-60, T 38.5-40.0 C, SpO2 ≥95%
FERRET     HR 180-250, RR 33-36, T 37.8-40.0 C, SpO2 ≥95%

PERFUSION INDEX (PI) FROM PLETH
PI < 0.4 %  : poor signal — SpO2 should not be trusted
PI 0.4-1.4  : weak — accept SpO2 if stable across 30+ s
PI 1.4-3.0  : adequate
PI > 3.0    : strong, room for improvement only with motion artifact

ECG GAIN BANDS
The AM4100 reports raw int8 deviation; the app converts to ~4 uV/LSB.
QRS amplitudes typically 0.5-3 mV in dogs/cats; values < 0.2 mV with
high signal_quality may indicate poor lead contact rather than low
voltage QRS.`;

export interface SystemBlocks {
  system: Array<{
    type: "text";
    text: string;
    cache_control?: { type: "ephemeral"; ttl?: "5m" | "1h" };
  }>;
}

export function buildSystemBlocks(): SystemBlocks {
  return {
    system: [
      {
        type: "text",
        text: SYSTEM_PROMPT,
      },
      {
        type: "text",
        text: SPECIES_REFERENCE,
        // Cache the prefix (system + species reference + tools render
        // here) for an hour. A vet clinic generates many insights per
        // hour during clinic hours — this is the place where caching
        // pays off the most.
        cache_control: { type: "ephemeral", ttl: "1h" },
      },
    ],
  };
}

export const OUTPUT_SCHEMA = {
  type: "object",
  properties: {
    summary: { type: "string" },
    findings: {
      type: "array",
      items: { type: "string" },
    },
    recommendations: {
      type: "array",
      items: { type: "string" },
    },
    urgency: {
      type: "string",
      enum: ["routine", "monitor", "urgent"],
    },
    thinking: { type: "string" },
  },
  required: ["summary", "findings", "recommendations", "urgency"],
  additionalProperties: false,
};

File: app/lib/ai/prompts.dart
/// Client-side helpers for building the JSON payload sent to the
/// `insight` Edge Function.
///
/// The actual prompt template + `cache_control` placement live
/// server-side in `backend/supabase/functions/insight/prompts.ts` so
/// the API key never leaves the server. This file only assembles the
/// **user-turn** content (per-session vitals, baselines, history
/// snippets) — the volatile suffix that intentionally sits *after*
/// the cached prefix.
library;

import '../data/models/insight.dart';
import '../data/models/pet.dart';
import '../data/models/session.dart';
import '../signal/species_baselines.dart';

class InsightRequest {
  const InsightRequest({
    required this.pet,
    required this.session,
    required this.recentSessions,
    required this.locale,
  });

  final Pet pet;
  final MonitoringSession session;

  /// Up to ~30 days of prior session summaries for trend / baseline
  /// reasoning. Older sessions are dropped to keep the payload small.
  final List<MonitoringSession> recentSessions;

  /// IETF language tag for the response (e.g. `en`, `en-GB`, `de`,
  /// `fr`, `es`, `pt`, `zh-Hant`).
  final String locale;

  Map<String, Object?> toJson() => {
        'pet': {
          'id': pet.id,
          'name': pet.name,
          'species': pet.species.name,
          'breed': pet.breed,
          'sex': pet.sex.name,
          'weight_kg': pet.weightKg,
          'age_months': pet.ageMonths,
          'notes': pet.notes,
        },
        'baseline': _baselineJson(pet.baseline),
        'session': session.toJson(),
        'recent_sessions': recentSessions
            .where((s) => s.id != session.id && s.endedAt != null)
            .take(10)
            .map((s) => {
                  'id': s.id,
                  'started_at': s.startedAt.toIso8601String(),
                  'duration_minutes': s.duration?.inMinutes,
                  'summary': s.summary.toJson(),
                })
            .toList(),
        'locale': locale,
      };

  Map<String, Object?> _baselineJson(SpeciesBaseline b) => {
        'species': b.species.name,
        'size': b.size.name,
        'heart_rate_bpm': {'low': b.heartRateBpm.low, 'high': b.heartRateBpm.high},
        'respiration_rate_bpm': {
          'low': b.respirationRateBpm.low,
          'high': b.respirationRateBpm.high,
        },
        'temperature_celsius': {
          'low': b.temperatureCelsius.low,
          'high': b.temperatureCelsius.high,
        },
        'spo2_percent': {'low': b.spo2Percent.low, 'high': b.spo2Percent.high},
        'systolic_mmhg': {'low': b.systolicMmHg.low, 'high': b.systolicMmHg.high},
        'diastolic_mmhg': {
          'low': b.diastolicMmHg.low,
          'high': b.diastolicMmHg.high,
        },
        'mean_arterial_mmhg': {
          'low': b.meanArterialMmHg.low,
          'high': b.meanArterialMmHg.high,
        },
      };
}

class InsightResponse {
  const InsightResponse({
    required this.summary,
    required this.findings,
    required this.recommendations,
    required this.urgency,
    required this.thinking,
    required this.modelId,
    required this.inputTokens,
    required this.outputTokens,
    required this.cacheReadTokens,
  });

  final String summary;
  final List<String> findings;
  final List<String> recommendations;
  final InsightUrgency urgency;
  final String thinking;
  final String modelId;
  final int inputTokens;
  final int outputTokens;
  final int cacheReadTokens;

  factory InsightResponse.fromJson(Map<String, Object?> json) =>
      InsightResponse(
        summary: (json['summary'] as String?) ?? '',
        findings: ((json['findings'] as List?) ?? const [])
            .map((e) => e.toString())
            .toList(),
        recommendations: ((json['recommendations'] as List?) ?? const [])
            .map((e) => e.toString())
            .toList(),
        urgency: InsightUrgency.values.byName(
          (json['urgency'] as String?) ?? 'routine',
        ),
        thinking: (json['thinking'] as String?) ?? '',
        modelId: (json['model'] as String?) ?? '',
        inputTokens: (json['usage'] as Map?)?['input_tokens'] as int? ?? 0,
        outputTokens: (json['usage'] as Map?)?['output_tokens'] as int? ?? 0,
        cacheReadTokens:
            (json['usage'] as Map?)?['cache_read_input_tokens'] as int? ?? 0,
      );
}
